Fix handling of v2 hashes

bernt-matthias opened the pull request:
Added tests first for the mulled container resolvers here https://github.com/galaxyproject/galaxy/pull/18522/commits/5b3c4bddd38d88bd195176be9f20f3e81d5336df

Then fixed (and more tests) here: https://github.com/galaxyproject/galaxy/pull/18522/commits/e7bca837f08345980f240f633921dbcce174345e

See details in https://github.com/galaxyproject/galaxy/pull/18522/commits/e7bca837f08345980f240f633921dbcce174345e

I'm tempted to backport this to 24.1 in order to make this effective in the multi-package-container repo.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

bernt-matthias committed:
add tests for mulled v2 hashes

to verify if the build is considered as part of the version or not

bernt-matthias committed:
fix mulled v2 hash generation from strings

the build information that may be contained in requirement versions is considered
as part of the version in some places of the code

- mulled container resolvers
- mulled-build-tool
- [planemo](https://github.com/galaxyproject/planemo/blob/b3e12a334e3d358ce30bf2c8c3f8d5a7a7e08136/planemo/conda.py#L116)

in others it isn't:

- mulled-hash
- mulled-build-files
- mulled_build

which are the files that use the `target_str_to_targets` function.

With this change the build info in version strings is always considered
as part of the version.

`mulled-build-files` is used in the multi-package-containers repo which
led to wrong hashes in up to approx 50 tools (check with `grep "=[^,]*=" combinations/* | wc -l`).

mvdbeek commented on test/functional/tools/mulled_example_multi_2.xml:
```diff
@@ -0,0 +1,79 @@
+<!-- same as mulled_example_multi_1, but requirements use builds
+     used to show that this does not yield the same mulled v2 hash,
+     i.e. the build is considered as part of version -->
+<tool id="mulled_example_multi_2" name="mulled_example_multi_2" version="0.1.0" license="AFL-3.0">
```
The unit tests cover this already, I would prefer that we don't add this. In the end this is just a bug in mulled_build_tool.py which we don't use in Galaxy.

bernt-matthias replied:
> The unit tests cover this already, I would prefer that we don't add this. In the end this is just a bug in mulled_build_tool.py which we don't use in Galaxy.

I can remove it. The idea was to make sure that the parser really gives a version string with the build (maybe this test should be added somewhere) .. but we would realize very quickly if something goes wrong here anyway.

bernt-matthias replied:
Removed the test.

bernt-matthias committed:
remove integration test

mvdbeek reviewed (changes requested):
Wait, this can't be right, why did you remove the build parsing ?

bernt-matthias committed:
fix linter errors

bernt-matthias committed:
resore test tool

needed for unit test

mvdbeek commented on lib/galaxy/tool_util/deps/mulled/mulled_build.py:
```diff
@@ -531,12 +531,7 @@ def target_str_to_targets(targets_raw):
-            elif "--" in version:
-                version, build = version.split("--")
-            target = build_target(package_name, version, build)
+            target = build_target(package_name, version)
```
I can't see how that won't break on:
```
combinations/hash.tsv
79:bedtools=2.17.0--0,python=2.7.13	bgruening/busybox-bash:0.1	0
94:picard=2.22.8--0,bwa=0.7.17	bgruening/busybox-bash:0.1	0
95:whatshap=1.0--py37h9a982cc_1,tabix=0.2.6--ha92aebf_0	bgruening/busybox-bash:0.1	0
```
Was there any reason not to use my suggestion ?

bernt-matthias replied:
Indeed we would change image names, but the new image names would be the ones which the container resolvers are looking for (in the current implementation).

mvdbeek replied:
It won't just change the image name, it will not install the correct build.

bernt-matthias replied:
Now I got it. `requirements_to_mulled_targets` also affects the names that the mulled resolvers expect: https://github.com/galaxyproject/galaxy/blob/a7fb5711431df5995b65330b5c99ac0503e8eec9/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L820C12-L820C42

I will try to finish this next week.

Sorry for the noise and thanks for double/triple checking.

bernt-matthias committed:
extend test to show that build information is still passed on

bernt-matthias committed:
Revert "remove integration test"

This reverts commit 565e228df26efe309af7be408f45e81ac3f323fc.

bernt-matthias committed:
Revert "fix mulled v2 hash generation from strings"

This reverts commit e7bca837f08345980f240f633921dbcce174345e.

bernt-matthias committed:
keep test

bernt-matthias committed:
fix .. the other way round

bernt-matthias committed:
fix tests

needs to be the same hash in all places and independent of given/absent
build in the version.

bernt-matthias commented on lib/galaxy/tool_util/deps/mulled/util.py:
```diff
@@ -335,7 +335,14 @@ def v2_image_name(
     'mulled-v2-fe8faa35dbf6dc65a0f7f5d4ea12e31a79f73e40'
+    >>> targets_version_with_build = [build_target("samtools", version="1.3.1=h9071d68_10"), build_target("bedtools", version="2.26.0=0")]
+    >>> v2_image_name(targets_version_with_build)
+    'mulled-v2-8186960447c5cb2faa697666dc1e6d919ad23f3e:8e86df67d257ce6494ae12b2c60e1b94025ea529'
```
```suggestion
```

bernt-matthias committed:
Update lib/galaxy/tool_util/deps/mulled/util.py

mvdbeek commented on lib/galaxy/tool_util/deps/mulled/mulled_build_tool.py:
```diff
@@ -20,13 +20,31 @@
+    >>> import doctest
+    >>> doctest.ELLIPSIS_MARKER = '-ignore-'
+    >>> import argparse
+    >>> _mulled_build_tool("test/functional/tools/mulled_example_multi_1.xml", argparse.Namespace(dry_run=True, base_image="does-not-matter-here-but-test-is-fast", command="build", verbose=True, involucro_path="./involucro"))  # doctest: +ELLIPSIS
```
This shouldn't be a doctest, and we don't want to run this with the general unit tests, just the mulled unit tests please.

mvdbeek reviewed (approved):
Thanks, I think this looks good to me now!

bernt-matthias commented on lib/galaxy/tool_util/deps/mulled/mulled_build_tool.py:
```diff
@@ -20,13 +20,31 @@
+    -ignore- TARGETS=samtools=1.3.1,bedtools=2.26.0 -ignore- REPO=quay.io/biocontainers/mulled-v2-8186960447c5cb2faa697666dc1e6d919ad23f3e:a6419f25efff953fc505dbd5ee734856180bb619-0 -ignore-
+    >>> _mulled_build_tool("test/functional/tools/mulled_example_multi_2.xml", argparse.Namespace(dry_run=True, base_image="does-not-matter-here-but-test-is-fast", command="build", verbose=True, involucro_path="./involucro"))  # doctest: +ELLIPSIS
+    -ignore- TARGETS=samtools=1.3.1=h9071d68_10,bedtools=2.26.0=0 -ignore- REPO=quay.io/biocontainers/mulled-v2-8186960447c5cb2faa697666dc1e6d919ad23f3e:a6419f25efff953fc505dbd5ee734856180bb619-0 -ignore-
+    """
```
```suggestion
```

bernt-matthias committed:
Remove doctest

mvdbeek commented on test/integration/test_container_resolvers.py:
```diff
@@ -480,6 +480,11 @@ class MulledTestCase:
 
 
+class MulledTestCaseWithBuildInfo:
+    tool_id = "mulled_example_multi_2"
```
How does this work if you don't add the tool to the sample tool config file ?

bernt-matthias replied:
Good point. I forgot to add an actual test case.

bernt-matthias committed:
enable integration test

mvdbeek committed:
Fix linting

bernt-matthias committed:
remove integration test

bernt-matthias committed:
fix bug for empty requirements

an empty set of requirements gives an empty requirement string which
was wrongly parsed to an CondaTarget with empty package name

This triggered a `conda search ''` which took so much memory that other
processes crashed.

The fix is to parse an empty list of CondaTargets for an empty target
string. In addition we check for empty package name during the creation
of the CondaTarget (to make this more easily detectable in the future)
